README.md
# `convert_extensions`

> Common convert methods exposed as static extensions method from the Dart core library and the convert package. Base64, ASCII, UTF8, Latin1, JSON, and more.

The `convert_extensions` package is a collection of [extension methods](https://dart.dev/guides/language/extension-methods)
based on the [`dart:convert` core library](https://api.dart.dev/stable/dart-convert/dart-convert-library.html),
the [`package:convert` package by the Dart team](https://pub.dev/packages/convert),
and other conversion methods that are found in the Dart language.

These extension methods make it easier to read and write code that uses conversions as it hides the conversion complexity behind expressive getters and methods.

Encode or decode with UTF8, Base64, JSON (and many more) with a simple and expressive API.

[![Continuous Integration](https://github.com/dartsidedev/convert_extensions/workflows/Continuous%20Integration/badge.svg?branch=main)](https://github.com/dartsidedev/convert_extensions/actions) [![codecov](https://codecov.io/gh/dartsidedev/convert_extensions/branch/main/graph/badge.svg)](https://codecov.io/gh/dartsidedev/convert_extensions) [![convert_extensions](https://img.shields.io/pub/v/convert_extensions?label=convert_extensions&logo=dart)](https://pub.dev/packages/convert_extensions 'See convert_extensions package info on pub.dev') [![Published by dartside.dev](https://img.shields.io/static/v1?label=Published%20by&message=dartside.dev&logo=dart&logoWidth=30&color=40C4FF&labelColor=1d599b&labelWidth=100)](https://pub.dev/publishers/dartside.dev/packages) [![GitHub Stars Count](https://img.shields.io/github/stars/dartsidedev/convert_extensions?logo=github)](https://github.com/dartsidedev/convert_extensions 'Star me on GitHub!')

## Important links

* [Read the source code and **star the repo** on GitHub](https://github.com/dartsidedev/convert_extensions)
* [Open an issue on GitHub](https://github.com/dartsidedev/convert_extensions/issues)
* [See package on pub.dev](https://pub.dev/packages/convert_extensions)
* [Read the docs on pub.dev](https://pub.dev/documentation/convert_extensions/latest/)

**If you enjoy using this package, a thumbs-up on [pub.dev](https://pub.dev/packages/convert_extensions) would be highly appreciated! 👍💙.**

## Resources

* [`dart:convert` core library](https://api.dart.dev/stable/dart-convert/dart-convert-library.html) 
* [`package:convert` package by the Dart team](https://pub.dev/packages/convert) 

## Motivation

Encoding and decoding directly with the Dart core library (`dart:convert` and others) and the `package:convert`
by the Dart team can be clunky, and it's sometimes hard to remember how to use those APIs.

Via extension methods, this package makes it easier to convert via encoders and decoders.

Keep in mind, that this package only provides static extension methods, and the actual conversion is still done via the
the Dart core library (`dart:convert` and others) and the `package:convert` package by the Dart team.

## Usage

Just import the package, and the static extension methods will be added.

```dart
import 'package:convert_extensions/convert_extensions.dart';

void main() {
  // ASCII
  print([99, 111, 110, 118, 101, 114, 116].asciiDecoded);
  print('convert'.asciiEncoded);
  
  // Base64
  print([99, 111, 110, 118, 101, 114, 116].base64Encoded);
  print('Y29udmVydA=='.base64Decoded);
  
  // JSON
  print('{"convert":true}'.jsonDecoded);
  print({'convert': true}.jsonEncoded);
  
  // UTF8
  print('convert'.utf8Encoded);
  print([99, 111, 110, 118, 101, 114, 116].utf8Decoded);
  
  // URI
  print('convert%20this'.uriDecoded);
  print('convert this'.uriEncoded);
  
  // URI query component
  print('convert+this'.uriQueryComponentDecoded);
  print('convert this'.uriQueryComponentEncoded);
  
  // HEX
  print('636f6e76657274'.hexDecoded);
  print([99, 111, 110, 118, 101, 114, 116].hexEncoded);
  
  // Percent
  print('hi/+'.percentDecoded);
  print([104, 105, 47, 43].percentEncoded);
}
```

You can also combine these getters. As an example, this is how you would create a basic auth header:

```dart
void main() {
  const username = 'johndoe';
  const password = r'p@$$w0rd';
  
  final key = '$username:$password'.utf8Encoded.base64Encoded;
  
  print('Basic $key');
}
```

For all extension methods, see the [API reference](https://pub.dev/documentation/convert_extensions/latest/).

You can see more [usage examples in the `test` folder](https://github.com/dartsidedev/convert_extensions/tree/main/test).


CHANGELOG.md
## 1.0.0

- Docs finished
- Examples improved
- Misnamed library fixed (`core_misc`)
- 1.0.0+1, 1.0.0+2 was testing the CI, no code changes

## 0.1.0

- Initial version.


test/core_misc_test.dart
import 'package:convert_extensions/core_misc.dart';
import 'package:test/test.dart';

void main() {
  test('uriDecoded', () {
    expect('convert%20this'.uriDecoded, 'convert this');
  });

  test('uriEncoded', () {
    expect('convert this'.uriEncoded, 'convert%20this');
  });

  test('uriComponentEncoded', () {
    expect('convert%20this'.uriComponentDecoded, 'convert this');
  });

  test('uriComponentEncoded', () {
    expect('convert this'.uriComponentEncoded, 'convert%20this');
  });

  test('uriQueryComponentDecoded', () {
    expect('convert+this'.uriQueryComponentDecoded, 'convert this');
  });

  test('uriQueryComponentEncoded', () {
    expect('convert this'.uriQueryComponentEncoded, 'convert+this');
  });
}


test/core_convert_test.dart
import 'dart:convert';

import 'package:convert_extensions/core_convert.dart';
import 'package:test/test.dart';

void main() {
  test('asciiDecoded', () {
    expect([99, 111, 110, 118, 101, 114, 116].asciiDecoded, 'convert');
  });

  test('asciiEncoded', () {
    expect('convert'.asciiEncoded, [99, 111, 110, 118, 101, 114, 116]);
    expect(() => 'cönvert'.asciiEncoded, throwsArgumentError);
  });

  test('base64Encoded', () {
    expect([99, 111, 110, 118, 101, 114, 116].base64Encoded, 'Y29udmVydA==');
  });

  test('base64Decoded', () {
    expect('Y29udmVydA=='.base64Decoded, [99, 111, 110, 118, 101, 114, 116]);
  });

  test('base64UrlEncoded', () {
    expect([99, 111, 110, 118, 101, 114, 116].base64UrlEncoded, 'Y29udmVydA==');
  });

  test('base64UrlDecoded', () {
    expect('Y29udmVydA=='.base64UrlDecoded, [99, 111, 110, 118, 101, 114, 116]);
  });

  test('base64Encoded (with UTF8)', () {
    expect('convert'.utf8Encoded.base64Encoded, 'Y29udmVydA==');
  });

  test('base64Decoded (with UTF8)', () {
    expect('Y29udmVydA=='.base64Decoded.utf8Decoded, 'convert');
  });

  test('base64Encoded and base64UrlEncoded (with UTF8)', () {
    expect('>">'.utf8Encoded.base64Encoded, 'PiI+');
    expect('>">'.utf8Encoded.base64UrlEncoded, 'PiI-');
  });

  test('htmlEscaped', () {
    expect(
      'import "dart:convert"; // << Import & name'.htmlEscaped,
      'import &quot;dart:convert&quot;; &#47;&#47; &lt;&lt; Import &amp; name',
    );
  });

  test('htmlEscape', () {
    expect(
      'import "dart:convert"; // << Import & name'.htmlEscape(
        HtmlEscapeMode(
          escapeSlash: false,
          escapeQuot: true,
          escapeLtGt: true,
        ),
      ),
      'import &quot;dart:convert&quot;; // &lt;&lt; Import &amp; name',
    );
  });

  test('jsonDecoded', () {
    expect('{"convert":true}'.jsonDecoded, {'convert': true});
  });

  test('jsonEncoded', () {
    expect({'convert': true}.jsonEncoded, '{"convert":true}');
  });

  test('jsonEncode', () {
    expect(
      {'convert': true}.jsonEncode(indent: '  '),
      '{\n  "convert": true\n}',
    );
  });

  test('jsonUtf8Encoded', () {
    expect(
      {'convert': 1}.jsonUtf8Encoded,
      [123, 34, 99, 111, 110, 118, 101, 114, 116, 34, 58, 49, 125],
    );
  });

  test('latin1Encoded', () {
    expect('convert'.latin1Encoded, [99, 111, 110, 118, 101, 114, 116]);
    expect('cönvert'.latin1Encoded, [99, 246, 110, 118, 101, 114, 116]);
  });

  test('latin1Decoded', () {
    expect([99, 111, 110, 118, 101, 114, 116].latin1Decoded, 'convert');
    expect([99, 246, 110, 118, 101, 114, 116].latin1Decoded, 'cönvert');
  });

  test('utf8Encoded', () {
    expect('convert'.utf8Encoded, [99, 111, 110, 118, 101, 114, 116]);
    expect('cönvert'.utf8Encoded, [99, 195, 182, 110, 118, 101, 114, 116]);
  });

  test('utf8Decoded', () {
    expect([99, 111, 110, 118, 101, 114, 116].utf8Decoded, 'convert');
    expect([99, 195, 182, 110, 118, 101, 114, 116].utf8Decoded, 'cönvert');
  });
}


test/package_convert_test.dart
import 'package:convert_extensions/package_convert.dart';
import 'package:test/test.dart';

void main() {
  test('hexEncoded', () {
    expect('636f6e76657274'.hexDecoded, [99, 111, 110, 118, 101, 114, 116]);
  });

  test('hexDecoded', () {
    expect([99, 111, 110, 118, 101, 114, 116].hexEncoded, '636f6e76657274');
  });

  test('percentDecoded', () {
    expect('hi%2F%2B'.percentDecoded, [104, 105, 47, 43]);
    expect('hi/+'.percentDecoded, [104, 105, 47, 43]);
  });

  test('percentEncoded', () {
    expect([104, 105, 47, 43].percentEncoded, 'hi%2F%2B');
  });

  // These tests could be improved further, but remember, this package
  // and its extensions are only using the Dart team's convert package
  // so extensive testing of encoding/decoding is not necessary
  // in this package.

  test('latin2Encoded', () {
    expect('Hello'.latin2Encoded, [72, 101, 108, 108, 111]);
    expect(
      'Árvíztűrő tükörfúrógép'.latin2Encoded,
      [
        193,
        114,
        118,
        237,
        122,
        116,
        251,
        114,
        245,
        32,
        116,
        252,
        107,
        246,
        114,
        102,
        250,
        114,
        243,
        103,
        233,
        112,
      ],
    );
  });

  test('latin2Decoded', () {
    expect([72, 101, 108, 108, 111].latin2Decoded, 'Hello');
  });

  test('latin3Encoded', () {
    expect('Hello'.latin3Encoded, [72, 101, 108, 108, 111]);
    expect(
      () => 'Árvíztűrő tükörfúrógép'.latin3Encoded,
      throwsFormatException,
    );
  });

  test('latin3Decoded', () {
    expect([72, 101, 108, 108, 111].latin3Decoded, 'Hello');
  });

  test('latin4Encoded', () {
    expect('Hello'.latin4Encoded, [72, 101, 108, 108, 111]);
    expect(
      () => 'Árvíztűrő tükörfúrógép'.latin4Encoded,
      throwsFormatException,
    );
  });

  test('latin4Decoded', () {
    expect([72, 101, 108, 108, 111].latin4Decoded, 'Hello');
  });

  test('latin5Encoded', () {
    expect('Hello'.latin5Encoded, [72, 101, 108, 108, 111]);
    expect(
      () => 'Árvíztűrő tükörfúrógép'.latin5Encoded,
      throwsFormatException,
    );
  });

  test('latin5Decoded', () {
    expect([72, 101, 108, 108, 111].latin5Decoded, 'Hello');
  });

  test('latin6Encoded', () {
    expect('Hello'.latin6Encoded, [72, 101, 108, 108, 111]);
    expect(
      () => 'Árvíztűrő tükörfúrógép'.latin6Encoded,
      throwsFormatException,
    );
  });

  test('latin6Decoded', () {
    expect([72, 101, 108, 108, 111].latin6Decoded, 'Hello');
  });

  test('latin7Encoded', () {
    expect('Hello'.latin7Encoded, [72, 101, 108, 108, 111]);
    expect(
      () => 'Árvíztűrő tükörfúrógép'.latin7Encoded,
      throwsFormatException,
    );
  });

  test('latin7Decoded', () {
    expect([72, 101, 108, 108, 111].latin7Decoded, 'Hello');
  });

  test('latin8Encoded', () {
    expect('Hello'.latin8Encoded, [72, 101, 108, 108, 111]);
    expect(
      () => 'Árvíztűrő tükörfúrógép'.latin8Encoded,
      throwsFormatException,
    );
  });

  test('latin8Decoded', () {
    expect([72, 101, 108, 108, 111].latin8Decoded, 'Hello');
  });

  test('latin9Encoded', () {
    expect('Hello'.latin9Encoded, [72, 101, 108, 108, 111]);
    expect(
      () => 'Árvíztűrő tükörfúrógép'.latin9Encoded,
      throwsFormatException,
    );
  });

  test('latin9Decoded', () {
    expect([72, 101, 108, 108, 111].latin9Decoded, 'Hello');
  });

  test('latin10Encoded', () {
    expect('Hello'.latin10Encoded, [72, 101, 108, 108, 111]);
    expect(
      'Árvíztűrő tükörfúrógép'.latin10Encoded,
      [
        193,
        114,
        118,
        237,
        122,
        116,
        248,
        114,
        245,
        32,
        116,
        252,
        107,
        246,
        114,
        102,
        250,
        114,
        243,
        103,
        233,
        112
      ],
    );
  });

  test('latin10Decoded', () {
    expect([72, 101, 108, 108, 111].latin10Decoded, 'Hello');
  });

  test('latinArabicEncoded', () {
    expect('Hello'.latinArabicEncoded, [72, 101, 108, 108, 111]);
    expect(
      () => 'Árvíztűrő tükörfúrógép'.latinArabicEncoded,
      throwsFormatException,
    );
  });

  test('latinArabicDecoded', () {
    expect([72, 101, 108, 108, 111].latinArabicDecoded, 'Hello');
  });

  test('latinCyrillicEncoded', () {
    expect('Hello'.latinCyrillicEncoded, [72, 101, 108, 108, 111]);
    expect(
      () => 'Árvíztűrő tükörfúrógép'.latinCyrillicEncoded,
      throwsFormatException,
    );
  });

  test('latinCyrillicDecoded', () {
    expect([72, 101, 108, 108, 111].latinCyrillicDecoded, 'Hello');
  });

  test('latinGreekEncoded', () {
    expect('Hello'.latinGreekEncoded, [72, 101, 108, 108, 111]);
    expect(
      () => 'Árvíztűrő tükörfúrógép'.latinGreekEncoded,
      throwsFormatException,
    );
  });

  test('latinGreekDecoded', () {
    expect([72, 101, 108, 108, 111].latinGreekDecoded, 'Hello');
  });

  test('latinHebrewEncoded', () {
    expect('Hello'.latinHebrewEncoded, [72, 101, 108, 108, 111]);
    expect(
      () => 'Árvíztűrő tükörfúrógép'.latinHebrewEncoded,
      throwsFormatException,
    );
  });

  test('latinHebrewDecoded', () {
    expect([72, 101, 108, 108, 111].latinHebrewDecoded, 'Hello');
  });

  test('latinThaiEncoded', () {
    expect('Hello'.latinThaiEncoded, [72, 101, 108, 108, 111]);
    expect(
      () => 'Árvíztűrő tükörfúrógép'.latinThaiEncoded,
      throwsFormatException,
    );
  });

  test('latinThaiDecoded', () {
    expect([72, 101, 108, 108, 111].latinThaiDecoded, 'Hello');
  });
}


lib/core_misc.dart
/// A collection of static extension methods that expose common conversion
/// methods from the Dart core library (expect those from `dart:convert`).
library core_misc;

extension UriDecoded on String {
  String get uriDecoded {
    return Uri.decodeFull(this);
  }
}

extension UriEncoded on String {
  /// Converts with [Uri.encodeFull].
  String get uriEncoded {
    return Uri.encodeFull(this);
  }
}

extension UriComponentDecoded on String {
  /// Converts with [decodeComponent].
  String get uriComponentDecoded {
    return Uri.decodeComponent(this);
  }
}

extension UriComponentEncoded on String {
  /// Converts with [Uri.encodeComponent].
  String get uriComponentEncoded {
    return Uri.encodeComponent(this);
  }
}

extension UriQueryComponentDecoded on String {
  /// Converts with [Uri.decodeQueryComponent].
  String get uriQueryComponentDecoded {
    return Uri.decodeQueryComponent(this);
  }
}

extension UriQueryComponentEncoded on String {
  /// Converts with [Uri.encodeQueryComponent].
  String get uriQueryComponentEncoded {
    return Uri.encodeQueryComponent(this);
  }
}


lib/core_convert.dart
/// A collection of static extension methods that expose common conversion
/// methods from the Dart core library (`dart:convert`).
library core_convert;

import 'dart:convert';
import 'dart:typed_data';

extension AsciiDecodedExtension on List<int> {
  /// Converts to string with [ascii.decoder].
  String get asciiDecoded {
    return ascii.decoder.convert(this);
  }
}

extension AsciiEncodedExtension on String {
  /// Converts to ASCII (bytes) with [ascii.encoder].
  Uint8List get asciiEncoded {
    return ascii.encoder.convert(this);
  }
}

extension Base64DecodedExtension on String {
  /// Converts to bytes with [base64.decoder].
  Uint8List get base64Decoded {
    return base64.decoder.convert(this);
  }
}

extension Base64EncodedExtension on List<int> {
  /// Converts to string with [base64.encoder].
  String get base64Encoded {
    return base64.encoder.convert(this);
  }
}

extension Base64UrlDecodedExtension on String {
  /// Converts to bytes with [base64Url.decoder].
  Uint8List get base64UrlDecoded {
    return base64Url.decoder.convert(this);
  }
}

extension Base64UrlEncodedExtension on List<int> {
  /// Converts to string with [base64Url.encoder].
  String get base64UrlEncoded {
    return base64Url.encoder.convert(this);
  }
}

extension HtmlEscapedExtension on String {
  /// Converts to string with [htmlEscape].
  String get htmlEscaped {
    return htmlEscape.convert(this);
  }
}

extension HtmlEscapeExtension on String {
  /// Converts to string with [HtmlEscape] with [mode].
  String htmlEscape(HtmlEscapeMode mode) {
    return HtmlEscape(mode).convert(this);
  }
}

extension JsonDecodedExtension on String {
  /// Converts to JSON with [json.decoder].
  dynamic get jsonDecoded {
    return json.decoder.convert(this);
  }
}

extension JsonEncodedExtension on Object? {
  /// Converts to JSON string with [json.encoder].
  String get jsonEncoded {
    return json.encoder.convert(this);
  }
}

extension JsonEncodeExtension on Object? {
  /// Converts to JSON string with [indent] and [toEncodable].
  String jsonEncode({
    String? indent,
    Object? Function(dynamic object)? toEncodable,
  }) {
    return JsonEncoder.withIndent(indent, toEncodable).convert(this);
  }
}

extension JsonUtf8EncodedExtension on Object? {
  /// Converts to bytes with [JsonUtf8Encoder].
  List<int> get jsonUtf8Encoded {
    return JsonUtf8Encoder().convert(this);
  }
}

extension Latin1DecodedExtension on List<int> {
  /// Converts to string with [latin1.decoder].
  String get latin1Decoded {
    return latin1.decoder.convert(this);
  }
}

extension Latin1EncodedExtension on String {
  /// Converts to bytes with [latin1.encoder].
  Uint8List get latin1Encoded {
    return latin1.encoder.convert(this);
  }
}

extension Utf8DecodedExtension on List<int> {
  /// Converts to string with [utf8.decoder].
  String get utf8Decoded {
    return utf8.decoder.convert(this);
  }
}

extension Utf8EncodedExtension on String {
  /// Converts to bytes with [utf8.encoder].
  Uint8List get utf8Encoded {
    return utf8.encoder.convert(this);
  }
}


lib/package_convert.dart
/// A collection of static extension methods that expose common conversion
/// methods from the `package:convert` package by the Dart team.
library package_convert;

import 'dart:typed_data';

import 'package:convert/convert.dart';

extension HexDecodedExtension on String {
  /// Converts hexadecimal strings to byte arrays with [hex.decoder].
  List<int> get hexDecoded {
    return hex.decoder.convert(this);
  }
}

extension HexEncodedExtension on List<int> {
  /// Converts byte arrays to hexadecimal strings with [hex.encoder].
  String get hexEncoded {
    return hex.encoder.convert(this);
  }
}

extension Latin2EncodedExtension on String {
  /// Converts to bytes with [latin2.encoder].
  Uint8List get latin2Encoded {
    return latin2.encoder.convert(this);
  }
}

extension Latin2DecodedExtension on List<int> {
  /// Converts to string with [latin2.decoder].
  String get latin2Decoded {
    return latin2.decoder.convert(this);
  }
}

extension Latin3EncodedExtension on String {
  /// Converts to bytes with [latin3.encoder].
  Uint8List get latin3Encoded {
    return latin3.encoder.convert(this);
  }
}

extension Latin3DecodedExtension on List<int> {
  /// Converts to string with [latin3.decoder].
  String get latin3Decoded {
    return latin3.decoder.convert(this);
  }
}

extension Latin4EncodedExtension on String {
  /// Converts to bytes with [latin4.encoder].
  Uint8List get latin4Encoded {
    return latin4.encoder.convert(this);
  }
}

extension Latin4DecodedExtension on List<int> {
  /// Converts to string with [latin4.decoder].
  String get latin4Decoded {
    return latin4.decoder.convert(this);
  }
}

extension Latin5EncodedExtension on String {
  /// Converts to bytes with [latin5.encoder].
  Uint8List get latin5Encoded {
    return latin5.encoder.convert(this);
  }
}

extension Latin5DecodedExtension on List<int> {
  /// Converts to string with [latin5.decoder].
  String get latin5Decoded {
    return latin5.decoder.convert(this);
  }
}

extension Latin6EncodedExtension on String {
  /// Converts to bytes with [latin6.encoder].
  Uint8List get latin6Encoded {
    return latin6.encoder.convert(this);
  }
}

extension Latin6DecodedExtension on List<int> {
  /// Converts to string with [latin6.decoder].
  String get latin6Decoded {
    return latin6.decoder.convert(this);
  }
}

extension Latin7EncodedExtension on String {
  /// Converts to bytes with [latin7.encoder].
  Uint8List get latin7Encoded {
    return latin7.encoder.convert(this);
  }
}

extension Latin7DecodedExtension on List<int> {
  /// Converts to string with [latin7.decoder].
  String get latin7Decoded {
    return latin7.decoder.convert(this);
  }
}

extension Latin8EncodedExtension on String {
  /// Converts to bytes with [latin8.encoder].
  Uint8List get latin8Encoded {
    return latin8.encoder.convert(this);
  }
}

extension Latin8DecodedExtension on List<int> {
  /// Converts to string with [latin8.decoder].
  String get latin8Decoded {
    return latin8.decoder.convert(this);
  }
}

extension Latin9EncodedExtension on String {
  /// Converts to bytes with [latin9.encoder].
  Uint8List get latin9Encoded {
    return latin9.encoder.convert(this);
  }
}

extension Latin9DecodedExtension on List<int> {
  /// Converts to string with [latin9.decoder].
  String get latin9Decoded {
    return latin9.decoder.convert(this);
  }
}

extension Latin10EncodedExtension on String {
  /// Converts to bytes with [latin10.encoder].
  Uint8List get latin10Encoded {
    return latin10.encoder.convert(this);
  }
}

extension Latin10DecodedExtension on List<int> {
  /// Converts to string with [latin10.decoder].
  String get latin10Decoded {
    return latin10.decoder.convert(this);
  }
}

extension LatinArabicEncodedExtension on String {
  /// Converts to bytes with [latinArabic.encoder].
  Uint8List get latinArabicEncoded {
    return latinArabic.encoder.convert(this);
  }
}

extension LatinArabicDecodedExtension on List<int> {
  /// Converts to string with [latinArabic.decoder].
  String get latinArabicDecoded {
    return latinArabic.decoder.convert(this);
  }
}

extension LatinCyrillicEncodedExtension on String {
  /// Converts to bytes with [latinCyrillic.encoder].
  Uint8List get latinCyrillicEncoded {
    return latinCyrillic.encoder.convert(this);
  }
}

extension LatinCyrillicDecodedExtension on List<int> {
  /// Converts to string with [latinCyrillic.decoder].
  String get latinCyrillicDecoded {
    return latinCyrillic.decoder.convert(this);
  }
}

extension LatinGreekEncodedExtension on String {
  /// Converts to bytes with [latinGreek.encoder].
  Uint8List get latinGreekEncoded {
    return latinGreek.encoder.convert(this);
  }
}

extension LatinGreekDecodedExtension on List<int> {
  /// Converts to string with [latinGreek.decoder].
  String get latinGreekDecoded {
    return latinGreek.decoder.convert(this);
  }
}

extension LatinHebrewEncodedExtension on String {
  /// Converts to bytes with [latinHebrew.encoder].
  Uint8List get latinHebrewEncoded {
    return latinHebrew.encoder.convert(this);
  }
}

extension LatinHebrewDecodedExtension on List<int> {
  /// Converts to string with [latinHebrew.decoder].
  String get latinHebrewDecoded {
    return latinHebrew.decoder.convert(this);
  }
}

extension LatinThaiEncodedExtension on String {
  /// Converts to bytes with [latinThai.encoder].
  Uint8List get latinThaiEncoded {
    return latinThai.encoder.convert(this);
  }
}

extension LatinThaiDecodedExtension on List<int> {
  /// Converts to string with [latinThai.decoder].
  String get latinThaiDecoded {
    return latinThai.decoder.convert(this);
  }
}

extension PercentDecodedExtension on String {
  /// Converts to byte arrays with [percent.decoder].
  List<int> get percentDecoded {
    return percent.decoder.convert(this);
  }
}

extension PercentEncodedExtension on List<int> {
  /// Converts to percent-encoded strings with [percent.encoder].
  String get percentEncoded {
    return percent.encoder.convert(this);
  }
}


example/convert_extensions_example.dart
import 'package:convert_extensions/convert_extensions.dart';

void main() {
  // ASCII
  print([99, 111, 110, 118, 101, 114, 116].asciiDecoded);
  print('convert'.asciiEncoded);

  // Base64
  print([99, 111, 110, 118, 101, 114, 116].base64Encoded);
  print('Y29udmVydA=='.base64Decoded);

  // JSON
  print('{"convert":true}'.jsonDecoded);
  print({'convert': true}.jsonEncoded);

  // UTF8
  print('convert'.utf8Encoded);
  print([99, 111, 110, 118, 101, 114, 116].utf8Decoded);

  // URI
  print('convert%20this'.uriDecoded);
  print('convert this'.uriEncoded);

  // URI query component
  print('convert+this'.uriQueryComponentDecoded);
  print('convert this'.uriQueryComponentEncoded);

  // HEX
  print('636f6e76657274'.hexDecoded);
  print([99, 111, 110, 118, 101, 114, 116].hexEncoded);

  // Percent
  print('hi/+'.percentDecoded);
  print([104, 105, 47, 43].percentEncoded);

  // Basic Auth (combined)
  const username = 'johndoe';
  const password = r'p@$$w0rd';

  final key = '$username:$password'.utf8Encoded.base64Encoded;

  print('Basic $key');
}
